test/liquid/combinators/tags/for_test.exs

```diff
@@ -0,0 +1,217 @@
+    ]
+
+    Enum.each(tags, fn tag ->
+      test_combinator(tag, &Parser.for/1,  [
+        for: [
+          for_collection: [
+            variable_name: "item",
+            value: {:variable, ["array"]},
+            limit_param: [2]
+          ],
+          for_body: [],
+          else: []
+        ]
+      ])
+    end)
+  end
+
+  test "for tag: offset parameter" do
+    tags = [
+      "{% for item in array offset:2 %}{% else %}{% endfor %}",
+      "{%for item in array offset:2%}{%else%}{%endfor%}",
+      "{%     for     item    in     array  offset:2  %}{%   else    %}{%    endfor    %}"
+    ]
+
+    Enum.each(tags, fn tag ->
+      test_combinator(tag, &Parser.for/1, [
+        for: [
+          for_collection: [
+            variable_name: "item",
+            value: {:variable, ["array"]},
+            offset_param: [2]
+          ],
+          for_body: [],
+          else: []
+        ]
+      ])
+    end)
+  end
+
+  test "for tag: reversed parameter" do
+    tags = [
+      "{% for item in array reversed %}{% else %}{% endfor %}",
+      "{%for item in array reversed%}{%else%}{%endfor%}",
+      "{%     for     item    in     array  reversed  %}{%   else    %}{%    endfor    %}"
+    ]
+
+    Enum.each(tags, fn tag ->
+      test_combinator(tag, &Parser.for/1, [
+        for: [
+          for_collection: [
+            variable_name: "item",
+            value: {:variable, ["array"]},
+            reversed_param: []
+          ],
+          for_body: [],
+          else: []
+        ]
+      ])
+    end)
+  end
+
+  test "for tag: range parameter" do
+    tags = [
+      "{% for i in (1..10) %}{{ i }}{% endfor %}",
+      "{%for i in (1..10)%}{{ i }}{% endfor %}",
+      "{%     for     i     in     (1..10)      %}{{ i }}{%     endfor     %}"
+    ]
+
+    Enum.each(tags, fn tag ->
+      test_combinator(
+        tag,
+        &Parser.for/1,
+        for: [
+          for_collection: [
+            variable_name: "i",
+            range_value: [start: 1, end: 10]
+          ],
+          for_body: [variable: ["i"]]
+        ]
+      )
+    end)
+  end
+
+  test "for tag: range with variables" do
+    test_combinator(
+      "{% for i in (my_var..10) %}{{ i }}{% endfor %}",
+      &Parser.for/1,
+      for: [
+        for_collection: [
+          variable_name: "i",
+          range_value: [start: "my_var", end: 10]
+        ],
+        for_body: [variable: ["i"]]
+      ]
+    )
+  end
+
+  test "for tag: break tag" do
+    test_combinator(
+      "{% for i in (my_var..10) %}{{ i }}{% break %}{% endfor %}",
+      &Parser.for/1,
+      for: [
+        for_collection: [
+          variable_name: "i",
+          range_value: [start: "my_var", end: 10]
+        ],
+        for_body: [variable: ["i"], break: []]
+      ]
+    )
+  end
+
+  test "for tag: continue tag" do
+    test_combinator(
+      "{% for i in (1..my_var) %}{{ i }}{% continue %}{% endfor %}",
+      &Parser.for/1,
+      for: [
```

Comment:
Any reason why you decided to go for a keyword list (`{:for, <keyword list>}`) and not a three element tuple? (`{:for, {collection, body, else}}`)

Reply:
This is the standard way NimblerParse returns the parsed markup, as a list. We can change it with additional transformations but the `else` part is optional and we choose use list with not fixed size collections